tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/WeightedSnapshotWithExemplars.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.tritium.metrics.registry;
+
+import com.codahale.metrics.Snapshot;
+import com.codahale.metrics.WeightedSnapshot;
+import com.codahale.metrics.WeightedSnapshot.WeightedSample;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link WeightedSnapshot} with support for storing exemplar metadata for each sample.
+ */
+final class WeightedSnapshotWithExemplars extends Snapshot implements ExemplarsCapture {
+
+    /**
+     * A single sample item with value, weights and optional exemplar metadata.
+     */
+    static class WeightedSampleWithExemplar {
+
+        private final long value;
+        private final double weight;
+        private final @Nullable Object exemplarMetadata;
+
+        WeightedSampleWithExemplar(long value, double weight, @Nullable Object exemplarMetadata) {
+            this.value = value;
+            this.weight = weight;
+            this.exemplarMetadata = exemplarMetadata;
+        }
+
+        public long getValue() {
+            return value;
+        }
+
+        public double getWeight() {
+            return weight;
+        }
+
+        @Nullable
+        public Object getExemplarMetadata() {
+            return exemplarMetadata;
+        }
+    }
+
+    private final WeightedSnapshot weightedSnapshot;
+    private final ExemplarMetadataProvider<?> exemplarProvider;
+    private final List<LongExemplar<Object>> exemplarMetadatas;
+
+    /**
+     * Create a new {@link Snapshot} with the given values.
+     *
+     * @param values an unordered set of values in the reservoir
+     */
+    WeightedSnapshotWithExemplars(ExemplarMetadataProvider<?> provider, Collection<WeightedSampleWithExemplar> values) {
+        final List<WeightedSample> weightedSamples = new ArrayList<>();
+        this.exemplarMetadatas = new ArrayList<>();
+
+        values.forEach(v -> {
+            weightedSamples.add(new WeightedSample(v.value, v.weight));
+            if (v.exemplarMetadata != null) {
+                exemplarMetadatas.add(DefaultLongExemplar.of(v.exemplarMetadata, v.value));
+            }
+        });
+
```

Comment:
I'd suggest presizing `weightedSamples`, and only allocate the `ArrayList` when copying exemplar metadata

```suggestion
        final List<WeightedSample> weightedSamples = new ArrayList<>(values.size());
        List<LongExemplar<Object>> metadatas = null;

        for (WeightedSampleWithExemplar v : values) {
            weightedSamples.add(new WeightedSample(v.value, v.weight));
            if (v.exemplarMetadata != null) {
                if (metadatas == null) {
                    metadatas = new ArrayList<>();
                }
                metadatas.add(DefaultLongExemplar.of(v.exemplarMetadata, v.value));
            }
        }

        this.exemplarMetadatas = (metadatas == null) ? List.of() : metadatas;
```

Reply:
Perhaps a presized ImmutableList.builderWithExpectedSize to avoid mutable state

Reply:
Yeah presizing and allocating on demand (esp given we might end up with many cases of no collected exemplars) sounds reasonable to me.
We can use the builder with expected size for weightedSamples - but we just use it to feed the `WeightedSnapshot` class which copies it anyway, and we then don't keep a reference. I think the ImmutableList has more gains for the exemplarMetadatas list since we do keep a reference to that - but we don't have an expected size for that one so we should probably just use the standard ImmutableList.builder(). Changed both in any case for consistency, keeping the optimization for when we have no metadata: https://github.com/palantir/tritium/pull/1992/commits/466a5b2dcd92bd0ba90bb5b961cd7106a8badb4f